include/dali/core/geom/vec.h

```diff
@@ -0,0 +1,533 @@
+  DALI_HOST_DEV constexpr T *end() { return &v[N]; }
+  DALI_HOST_DEV constexpr const T *cend() const { return &v[N]; }
+  DALI_HOST_DEV constexpr const T *end() const { return &v[N]; }
+
+  /// @brief Calculates the sum of squares of components.
+  DALI_HOST_DEV constexpr auto length_square() const {
+    decltype(v[0]*v[0] + v[0]*v[0]) ret = v[0]*v[0];
+    for (size_t i = 1; i < N; i++)
+      ret += v[i]*v[i];
+    return ret;
+  }
+
+  /// @brief Calculates Euclidean length of the vector.
+  DALI_HOST_DEV inline auto length() const {
+#ifdef __CUDA_ARCH__
+    return sqrtf(length_square);
+#else
+  return std::sqrt(length_square());
+#endif
+  }
+  DALI_HOST_DEV inline vec normalized() const {
+    auto lsq = length_square();
+    return *this * rsqrt(lsq);
+  }
+
+  /// @brief Returns a copy. Doesn't promoto type to int.
+  DALI_HOST_DEV constexpr vec operator+() const { return *this; }
+
+  /// @brief Negates all components. Doesn't promote type to int.
+  DALI_HOST_DEV
+  inline vec operator-() const {
+    vec<N, T> ret;
+    for (size_t i = 0; i < N; i++) {
+      ret.v[i] = -v[i];
+    }
+    return ret;
+  }
+  DALI_HOST_DEV
+  inline vec operator~() const {
+    vec<N, T> ret;
+    for (size_t i = 0; i < N; i++) {
+      ret.v[i] = ~v[i];
+    }
+    return ret;
+  }
+
+  #define DEFINE_ASSIGN_VEC_OP(op)\
+  template <typename U>\
+  DALI_HOST_DEV vec &operator op(const vec<N, U> &rhs) {\
+    for (size_t i = 0; i < N; i++)\
+      v[i] op rhs[i];\
+    return *this;\
+  }\
+  template <typename U>\
+  DALI_HOST_DEV std::enable_if_t<!is_vec<U>::value, vec &> operator op(const U &rhs) {\
+    for (size_t i = 0; i < N; i++)\
+      v[i] op rhs;\
+    return *this;\
+  }
+
+  DEFINE_ASSIGN_VEC_OP(=)
+  DEFINE_ASSIGN_VEC_OP(+=)
+  DEFINE_ASSIGN_VEC_OP(-=)
+  DEFINE_ASSIGN_VEC_OP(*=)
+  DEFINE_ASSIGN_VEC_OP(/=)
+  DEFINE_ASSIGN_VEC_OP(%=)
+  DEFINE_ASSIGN_VEC_OP(&=)
+  DEFINE_ASSIGN_VEC_OP(|=)
+  DEFINE_ASSIGN_VEC_OP(^=)
+  DEFINE_ASSIGN_VEC_OP(<<=)
+  DEFINE_ASSIGN_VEC_OP(>>=)
+  #undef DEFINE_ASSIGN_VEC_OP
+};
+
+
+template <size_t N, typename T, typename U>
+DALI_HOST_DEV
+constexpr auto dot(const vec<N, T> &a, const vec<N, U> &b) {
+  decltype(a[0]*b[0] + a[0]*b[0]) ret = a[0]*b[0];
+  for (size_t i = 1; i < N; i++)
+    ret += a[i]*b[i];
+  return ret;
+}
+
+template <typename T, typename U>
+DALI_HOST_DEV
+constexpr auto cross(const vec<3, T> &a, const vec<3, U> &b) {
+  using R = decltype(a[0]*b[0] + a[0]*b[0]);
+  return vec<3, R>{
+    a.y * b.z - b.y * a.z,
+    a.z * b.x - b.z * a.x,
+    a.x * b.y - b.x * a.y
+  };
+}
+
+/// @brief Calculates `z` coordinate of a cross product of two 2D vectors
+template <typename T, typename U>
+DALI_HOST_DEV
+constexpr auto cross(const vec<2, T> &a, const vec<2, U> &b) {
+  return a.x * b.y - b.x * a.y;
+}
+
+#define DEFINE_ELEMENTIWSE_VEC_BIN_OP(op)\
+template <size_t N, typename T, typename U>\
+DALI_HOST_DEV inline auto operator op(const vec<N, T> &a, const vec<N, U> &b) {\
```

Comment:
This has unfortunate effect of `decltype(vec<4, int16_t>() + vec<4, int16_t>()) == decltype(vec<4, int32_t>)`. 
Is that desired bahaviour?

Reply:
Fixed. Now the promotion rules are as follows:
For vectors:
non-fp (op) fp -> fp
fp (op) non-fp -> fp
non-fp1 (op) non-fp2 -> error
fp1 (op) fp2 -> error
For vectors-scalar:
non-fp (op) fp -> fp
fp (op) non-fp -> fp
int_v (op) int_s -> int_s
fp_v (op) fp_s -> fp_v